bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/ExpireManager.java

```diff
@@ -0,0 +1,395 @@
+    }
+
+    private boolean isReadyToExpire(String itemName) {
+        Instant nextExpiry = itemExpireMap.get(itemName);
+        return (nextExpiry != null) && Instant.now().isAfter(nextExpiry);
+    }
+
+    private void expire(String itemName) {
+        itemExpireMap.remove(itemName); // disable expire trigger until next update or command
+        Optional<ExpireConfig> expireConfig = itemExpireConfig.get(itemName);
+
+        if (expireConfig != null && expireConfig.isPresent()) {
+            Command expireCommand = expireConfig.get().expireCommand;
+            State expireState = expireConfig.get().expireState;
+
+            if (expireCommand != null) {
+                logger.debug("Item {} received no command or update for {} - posting command '{}'", itemName,
+                        expireConfig.get().duration, expireCommand);
+                postCommand(itemName, expireCommand);
+            } else if (expireState != null) {
+                logger.debug("Item {} received no command or update for {} - posting state '{}'", itemName,
+                        expireConfig.get().duration, expireState);
+                postUpdate(itemName, expireState);
+            }
+        }
+    }
+
+    @Override
+    public Set<String> getSubscribedEventTypes() {
+        return SUBSCRIBED_EVENT_TYPES;
+    }
+
+    @Override
+    public @Nullable EventFilter getEventFilter() {
+        return null;
+    }
+
+    @Override
+    public void receive(Event event) {
+        if (!enabled) {
+            return;
+        }
+        if (event instanceof ItemStateEvent) {
+            receiveUpdate((ItemStateEvent) event);
+        } else if (event instanceof ItemCommandEvent) {
+            receiveCommand((ItemCommandEvent) event);
+        }
+    }
+
+    @Override
+    public void added(Item item) {
+        itemExpireConfig.remove(item.getName());
+    }
+
+    @Override
+    public void removed(Item item) {
+        itemExpireConfig.remove(item.getName());
+    }
+
+    @Override
+    public void updated(Item oldItem, Item item) {
+        itemExpireConfig.remove(item.getName());
+    }
+
+    class MetadataChangeListener implements RegistryChangeListener<Metadata> {
+
+        @Override
+        public void added(Metadata element) {
+            itemExpireConfig.remove(element.getUID().getItemName());
+        }
+
+        @Override
+        public void removed(Metadata element) {
+            itemExpireConfig.remove(element.getUID().getItemName());
+        }
+
+        @Override
+        public void updated(Metadata oldElement, Metadata element) {
+            itemExpireConfig.remove(element.getUID().getItemName());
+        }
+    }
+
+    class ExpireConfig {
+
+        protected static final String COMMAND_PREFIX = "command=";
+        protected static final String STATE_PREFIX = "state=";
+
+        protected final Pattern durationPattern = Pattern.compile("(?:([0-9]+)H)?\\s*(?:([0-9]+)M)?\\s*(?:([0-9]+)S)?",
```

Comment:
This can be `static final`.

Reply:
```
The field durationPattern cannot be declared static in a non-static inner type, unless initialized with a constant expression
```
As I don't want to move it to the parent class, I would keep it as is. Or do you think doing the pattern compilation is too expensive here?